table/block_based/block_based_table_builder.cc

```diff
@@ -684,6 +766,83 @@ struct BlockBasedTableBuilder::ParallelCompressionRep {
   }
 
   ~ParallelCompressionRep() { block_rep_pool.finish(); }
+
+  // Make a block prepared to be emitted to compression thread
+  BlockRep* PrepareBlock(CompressionType compression_type,
+                         const Slice* first_key_in_next_block,
+                         BlockBuilder* data_block) {
+    BlockRep* block_rep =
+        PrepareBlockInternal(compression_type, first_key_in_next_block);
+    data_block->SwapAndReset(*(block_rep->data));
+    block_rep->contents = *(block_rep->data);
+    std::swap(block_rep->keys, curr_block_keys);
+    curr_block_keys->Clear();
+    return block_rep;
+  }
+
+  BlockRep* PrepareBlock(CompressionType compression_type,
+                         const Slice* first_key_in_next_block,
+                         std::string* data_block,
+                         std::vector<std::string>* keys) {
+    BlockRep* block_rep =
+        PrepareBlockInternal(compression_type, first_key_in_next_block);
+    std::swap(*(block_rep->data), *data_block);
+    block_rep->contents = *(block_rep->data);
+    block_rep->keys->SwapAssign(*keys);
+    return block_rep;
+  }
+
+  // Emit a block to compression thread
+  void EmitBlock(BlockRep* block_rep) {
+    // Read out first_block here to avoid data race with BGWorkWriteRawBlock
+    bool tmp_first_block = first_block;
```

Comment:
copy & paste from a code analysis warning:

There may be a Lock Consistency Violation: Read/Write race. Non-private method rocksdb::BlockBasedTableBuilder::ParallelCompressionRep::EmitBlock reads without synchronization from this->first_block. Potentially races with write in method rocksdb::BlockBasedTableBuilder::ParallelCompressionRep::ReapBlock. (error trace: TV107924839)